### PR TITLE
[Work in progress] Cache CR

### PR DESCRIPTION
**Pull request by slaskawi**
https://github.com/infinispan/infinispan-operator/issues/224

This PR Includes:
- Migration to Go modules (Go `1.13`)
- CRD skeleton implementation generated by `Operator SDK v0.12.0-68-g7a5bb22a`.
- Example CR

Still to be done:
* [ ] Implementation
* [ ] Tests
* [ ] Documentation

Discussion points:
* Operator SDK used for Infinispan Operator is really old. We should probably consider an upgrade
* Operator SDK 1.12 supports only Go `1.13`. I did a first step generating Go Module files and removed Dep.
* Is API OK? Do you like YAML representation of Infinispan Cache Configuration?

### CODE REVIEW

**Comment by rigazilla** [comment]
Hey @slaskawi, thanks for the go modules migration. We really want it!
Can you create a separate PR? it would be great to have it merged soon

File: deploy/crds/infinispan.org_caches_crd.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                    {key,value} in the matchLabels map is equivalent to an element
+                    of matchExpressions, whose key field is "key", the operator is
+                    "In", and the values array contains only "value". The requirements
+                    are ANDed.
+                  type: object
+              type: object
+          required:
+          - cacheConfigurationSpec
+          type: object
+        status:
```
Comment (rigazilla) [inline deploy/crds/infinispan.org_caches_crd.yaml]: There's only the `conditions` field specified atm (statefulset?). We will provide connection info for clients also?

File: deploy/crds/infinispan.org_caches_crd.yaml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          description: 'Kind is a string value representing the REST resource this
+            object represents. Servers may infer this from the endpoint the client
+            submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds'
+          type: string
+        metadata:
+          type: object
+        spec:
+          description: CacheSpec defines the desired state of Cache
+          properties:
+            cacheConfigurationSpec:
```
Comment (rigazilla) [inline deploy/crds/infinispan.org_caches_crd.yaml]: `cacheConfigurationSpec` content is not specified. There will be a direct yaml->json mapping before calling the ispn rest api?

File: pkg/apis/infinispan/v1alpha1/cache_types.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	// Selector for looking up Infinispan Custom Resources.
+	// +kubebuilder:validation:Required
+	InfinispanSelector *metav1.LabelSelector `json:"infinispanSelector,omitempty"`
+	// A user-provided Cache configuration.
+	// +kubebuilder:validation:Required
+	CacheConfigurationSpec CacheConfigurationSpec `json:"cacheConfigurationSpec"`
+}
+
+type CacheConfigurationSpec interface {
+
```
Comment (rigazilla) [inline pkg/apis/infinispan/v1alpha1/cache_types.go]: Why is this an interface?

File: pkg/apis/infinispan/v1alpha1/cache_types.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// +k8s:openapi-gen=true
+type CacheStatus struct {
+	// Represents the latest available observations of a statefulset's current state.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []CacheCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,10,rep,name=conditions"`
+}
+
+// StatefulSetCondition describes the state of a statefulset at a certain point.
```
Comment (rigazilla) [inline pkg/apis/infinispan/v1alpha1/cache_types.go]: comment needs update

**Comment by slaskawi** [comment]
Let me close this for now...

**Pull request closed without merging**